src/merkle.rs

```diff
@@ -209,16 +309,228 @@ impl<T: Element, A: Algorithm<T>, K: Store<T>, U: Unsigned> MerkleTree<T, A, K,
+            _r: PhantomData,
+        })
+    }
+
+    /// Creates new compound merkle tree from a vector of merkle
+    /// trees.  The ordering of the trees is significant, as trees are
+    /// leaf indexed / addressable in the same sequence that they are
+    /// provided here.
+    pub fn from_trees(trees: Vec<MerkleTree<T, A, K, U>>) -> Result<MerkleTree<T, A, K, U, N, R>> {
+        ensure!(
+            N::to_usize() > 0,
+            "Cannot use from_trees if not constructing a structure with sub-trees"
+        );
+        ensure!(
+            trees.iter().all(|ref mt| mt.height() == trees[0].height()),
+            "All passed in trees must have the same height"
+        );
+        ensure!(
+            trees.iter().all(|ref mt| mt.len() == trees[0].len()),
+            "All passed in trees must have the same length"
+        );
+
+        let sub_tree_layer_nodes = N::to_usize();
+        ensure!(
+            trees.len() == sub_tree_layer_nodes,
+            "Length of trees MUST equal the number of sub tree layer nodes"
+        );
+
+        // If we are building a compound tree with no sub-trees,
+        // all properties revert to the single tree properties.
+        let (leafs, len, height, root) = if sub_tree_layer_nodes == 0 {
+            (
+                trees[0].leafs(),
+                trees[0].len(),
+                trees[0].height(),
+                trees[0].root(),
+            )
+        } else {
+            // Total number of leafs in the compound tree is the combined leafs total of all subtrees.
+            let leafs = trees.iter().fold(0, |leafs, mt| leafs + mt.leafs());
+            // Total length of the compound tree is the combined length of all subtrees plus the root.
+            let len = trees.iter().fold(0, |len, mt| len + mt.len()) + 1;
+            // Total height of the compound tree is the height of any of the sub-trees to top-layer plus root.
+            let height = trees[0].height() + 1;
+            // Calculate the compound root by hashing the top layer roots together.
+            let roots: Vec<T> = trees.iter().map(|x| x.root()).collect();
+            let root = A::default().multi_node(&roots, 1);
+
+            (leafs, len, height, root)
+        };
+
+        Ok(MerkleTree {
+            data: None,
+            leafs,
+            len,
+            height,
+            top_layer_nodes: R::to_usize(),
+            trees: None,
+            sub_tree_layer_nodes,
+            sub_trees: Some(trees),
+            root,
+            _t: PhantomData,
+            _a: PhantomData,
+            _u: PhantomData,
+            _n: PhantomData,
+            _r: PhantomData,
+        })
+    }
+
+    /// Creates new top layer merkle tree from a vector of merkle
+    /// trees with sub-trees.  The ordering of the trees is
+    /// significant, as trees are leaf indexed / addressable in the
+    /// same sequence that they are provided here.
+    pub fn from_sub_trees(
```

Comment:
this seems to be the exact same implementation as `from_trees` what am I missing?

Reply:
Should probably be simplified, but it's the assignment of `trees` vs `sub-trees` in the end construction.

Reply:
And the input type

Reply:
IIRC, the input type was the difficult to resolve part.

Reply:
I see it now, not sure of an easy way to combine them.